_posts/2024-04-23-kubernetes-horizontal-pod-autoscaler-guide.md

```diff
@@ -0,0 +1,164 @@
+
+For example Statsbomb can use HPA to handle increased traffic during a weekend when there are more games being played. The HPA can automatically scale up the number of web server pods to maintain performance, and scale down during off-peak hours to reduce costs. This dynamic adjustment helps ensure that the application consistently meets performance targets without manual intervention.
+
+In next sections I will briefly explain how HPA works and how to use this simple tool to ease your scaling decisions.
+
+## How does HPA work?
+
+First of all we need to make sure we understand the concepts of `request` and `limits` since they are fundamental to how resources are allocated and managed across the pods in a cluster.
+
+- **Requests**: This value specifies the amount of CPU or memory that Kubernetes guarantees to a pod. When a pod is scheduled, the Kubernetes scheduler uses this request value to decide on which node the pod can fit. So this number ensures the pod has the resources it needs to run.
+- **Limits**: This value specifies the maximum amount of CPU or memory that a pod can use. If a pod exceeds this limit, Kubernetes will throttle the pod or kill it. This is how k8s ensures that a single pod does not consume all the resources in a node.
+
+Imagine our deployment has the following setup:
+
+```yaml
+# deployment.yaml
+resources:
+  limits:
+    memory: 2000Mi
+    cpu: 1500m
+  requests:
+    memory: 1350Mi
+    cpu: 400m
+    
+```
+
+And our HPA is configured in the following manner:
+
+```yaml
+apiVersion: autoscaling/v2
+kind: HorizontalPodAutoscaler
+metadata:
+  name: autoscaler-name
+spec:
+  scaleTargetRef:
+    apiVersion: apps/v1
+    kind: Deployment
+    name: autoscaler-name
+  minReplicas: 1
+  maxReplicas: 3
+  metrics:
+  - type: Resource
+    resource:
+      name: cpu
+      target:
+        type: Utilization
+        averageUtilization: 90
+  - type: Resource
+    resource:
+      name: memory
+      target:
+        type: Utilization
+        averageUtilization: 90
+```
+
+So what does this mean? Well if you are a proper engineer what you would do is check the official docs and try to carefully understand this. But if you are like me... you probably would make some assumptions and hope for the best (for finally ending up reading the docs 😅). 
+
+My first guess was that this would configure HPA to to scale up the number of pods when memory or CPU usage exceeded 90%. However, I overlooked a crucial detail: the calculation also needs to include the current number of replicas. Here’s how HPA actually works:
+
+```
+desiredReplicas = ceil[currentReplicas * ( currentMetricValue / desiredMetricValue )]
+```
+
+Where:
+- `currentReplicas` is the number of replicas the deployment is currently running.
+- `currentMetricValue` is the current value of the metric we are monitoring (e.g. CPU usage).
+- `desiredMetricValue` is the target value for the metric we are monitoring (e.g. 90% requested CPU usage).
```

Comment:
I think it's important to clarify how this is computed. Is it
`desiredMetricValue = target.averageUtilization * requests (0.9*1350)`
or
`desiredMetricValue = target.averageUtilization * limits (0.9*2000)`

ChatGPT told me it was using limits, but this [post](https://pauldally.medium.com/horizontalpodautoscaler-uses-request-not-limit-to-determine-when-to-scale-97643d808997) says it's using requests.

Reply:
I will add that, it uses requests